---
title: "R_Package_Leow"
author: "ChiJingLeow"
date: "12/2/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Installation

Load in these libraries.
```{r}
library(ggplot2)
library(tidyverse)
library(devtools)
```

You might need to install these R packages for this package to run. 
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("edgeR")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("limma")
```

Load the packages.
```{r}
library("limma")
library("edgeR")
```

Now you can install my R package!
and load it...
```{r}
devtools::install_github("ChiJingLeow/R_Package_Leow")
library("RPackageLeow")
```

Download the data for this package. 

These data come from a Nature Cell Biology paper, EGF-mediated induction of Mcl-1 at the switch to lactation is essential for alvaelor cell survival (Fu et al., 2015).

The first file is the gene counts and the second file is the sample information.
```{r}
download.file("https://raw.githubusercontent.com/ChiJingLeow/R_Package_Leow/master/Data/GSE60450_LactationGenewiseCounts.txt", destfile = "/cloud/project/Data/Genedata.txt")

download.file("https://raw.githubusercontent.com/ChiJingLeow/R_Package_Leow/master/Data/SampleInfo_Corrected.txt", destfile = "/cloud/project/Data/Sampleinfo.txt")
```

Read the gene counts and sample info files using read.delim
```{r}
data <- read.delim("/cloud/project/Data/Genedata.txt")
sampleinfo <- read.delim("/cloud/project/Data/Sampleinfo.txt")
```

The gene counts data is in a table which includes GeneID, length, 12 samples, and the reads. 

Function 1: countdata_no_NA

The function takes in a dataset and removes potential NA values, geneID and length. 

It returns dataset with only counts.
```{r include=FALSE}
countdata_no_NA <- function(data) {
  if(is.data.frame(data) == FALSE) {
    return("not a dataframe")
  }
  product <- na.omit(data)
  if(product == data){
    print("no NA values found")
  }
  product <- data[,-(1:2)]
  return(product)
}

countdata <- countdata_no_NA(data)
countdata
```

The sample names are way too long...

Function 2: Short_name

The function takes in the dataset and returns with shortened sample names.
You should see the sample names like MCL1.DG rather than 30 characters in length.
```{r include=FALSE}
short_name <- function(data) {
  if(is.data.frame(data) == FALSE) {
    return("not a dataframe")
  }
  colnames(data) <- substr(colnames(data),start=1,stop=7)
  return(data)
  
}
countdata <- short_name(countdata)
countdata
```

Function 3: DGEobject

We can make a DGElist object from the package edgeR, "counts" and "samples" are stored in it.

This function takes in the dataset, creates a DGElist and adds the group category from sampleinfo to it.

You should see the group and library size are not stored in the DGElist.
```{r}
DGEobject <- function(countdata, sampleinfo) {
  if(is.data.frame(countdata) == FALSE) {
    return("not a dataframe")
  }
  if(is.data.frame(sampleinfo) == FALSE) {
    return("not a dataframe")
  }
  product <- DGEList(countdata)
  combined <- paste(sampleinfo$CellType, sampleinfo$Status, sep = "_")
  combined <- factor(combined) 
  product$samples$group <- combined
  return(product)
}
object1 <- DGEobject(countdata,sampleinfo)
object1
```

Function 4 :Filter_low_counts

Filtering out low counts (lowly expressed genes) is one of the first steps in analyzing gene expressions. 

The function filters out counts per million (cpm) that's lower than 0.5 and gives you a summary of the counts above the threshold.
```{r}
Filter_low_counts <- function(data) {
  if(is.data.frame(data) == FALSE) {
    return("not a dataframe")
  }
  data_cpm <- cpm(data)
  threshold <- data_cpm > 0.5
  threshold <- summary(threshold)
  return(threshold)
}
filtered <- Filter_low_counts(countdata)
filtered
```

To compare the library sizes from each group, we can use 

Function 5: plot_library_reads

The functions takes in the group and the library sieze. Red dots are the mean from each group.
```{r}
plot_library_reads <- function(group, lib_size) {
  if(is.numeric(lib_size) == FALSE) {
    return("not a numeric data")
  }
  ggplot(mapping = aes(x = group, y = lib_size)) +geom_point() +ggtitle("lib size from each group") +stat_summary(fun.y = mean, col ="red", fill = "red")
}

plot_library_reads(object1$samples$group , object1$samples$lib.size)
```
